File: apps/gateway/src/agent/secret-scan.ts
// 凭证检测与消毒（高精度模式）。只匹配高置信度凭证，尽量不误伤普通配置/散文。
// - redactSecrets：把命中片段替换成 [REDACTED:<type>]，用于「机器来源内容」出站 LLM 前消毒。
// - detectSecrets：只检测不改写，用于用户输入消息的泄露告警。

export interface SecretMatch {
  type: string;
}

interface SecretRule {
  type: string;
  regex: RegExp;
  // 用命中的捕获组重建消毒后文本（保留前缀，仅抹掉密值）。
  redact: (groups: string[]) => string;
}

const REDACTED = (type: string) => `[REDACTED:${type}]`;

// 注意：所有 regex 必须带 g 标志；replace 回调里据捕获组重建。
const RULES: SecretRule[] = [
  {
    type: 'private-key',
    regex: /-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z0-9 ]*PRIVATE KEY-----/g,
    redact: () => REDACTED('private-key'),
  },
  {
    // 含密码的连接串/URL：scheme://user:pass@host —— 仅抹掉 pass
    type: 'url-credential',
    regex: /\b([a-z][a-z0-9+.-]*:\/\/[^\s/:@]+:)([^\s/@]+)@/gi,
    redact: (g) => `${g[1]}${REDACTED('password')}@`,
  },
  {
    type: 'bearer-token',
    regex: /(Authorization:\s*Bearer\s+)([A-Za-z0-9._~+/-]+=*)/gi,
    redact: (g) => `${g[1]}${REDACTED('token')}`,
  },
  {
    // 已知前缀的 API token / access key
    type: 'api-token',
    regex:
      /(?:sk-[A-Za-z0-9_-]{16,}|gh[posru]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,}|xox[abprs]-[A-Za-z0-9-]{10,}|AKIA[0-9A-Z]{16}|ya29\.[A-Za-z0-9._-]{20,}|AIza[0-9A-Za-z_-]{30,}|glpat-[A-Za-z0-9_-]{18,})/g,
    redact: () => REDACTED('token'),
  },
  {
    // 网络设备带加密类型号的口令：`password 7 <hex>` / `secret 5 <hash>`
    type: 'device-secret',
    regex: /\b(password|secret)\s+([0-9])\s+(\S+)/gi,
    redact: (g) => `${g[1]} ${g[2]} ${REDACTED('device-secret')}`,
  },
  {
    // `enable secret <pass>`（可带类型号）
    type: 'device-secret',
    regex: /\b(enable\s+secret)\s+(?:[0-9]\s+)?(\S+)/gi,
    redact: (g) => `${g[1]} ${REDACTED('device-secret')}`,
  },
  {
    // SNMP community 串
    type: 'device-secret',
    regex: /\b(snmp-server\s+community)\s+(\S+)/gi,
    redact: (g) => `${g[1]} ${REDACTED('device-secret')}`,
  },
];

export function redactSecrets(input: string): { text: string; matches: SecretMatch[] } {
  if (!input) {
    return { text: input, matches: [] };
  }
  let text = input;
  const matches: SecretMatch[] = [];
  for (const rule of RULES) {
    text = text.replace(rule.regex, (...args) => {
      // args = [fullMatch, ...captureGroups, offset, string, (groups?)]
      // 去掉末尾的 offset/string（及可能的 named groups 对象），保留 fullMatch + 捕获组
      const groups = args.filter((arg) => typeof arg === 'string') as string[];
      // 最后一个 string 是整串 input，剔除它
      groups.pop();
      matches.push({ type: rule.type });
      return rule.redact(groups);
    });
  }
  return { text, matches };
}

export function detectSecrets(input: string): SecretMatch[] {
  return redactSecrets(input).matches;
}

export function hasSecret(input: string): boolean {
  return detectSecrets(input).length > 0;
}

File: apps/gateway/src-rust/agent/secret_scan.rs
use std::collections::BTreeSet;
use std::sync::LazyLock;

use regex::{Captures, Regex};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SecretMatch {
    pub kind: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct RedactedText {
    pub text: String,
    pub matches: Vec<SecretMatch>,
}

struct SecretPatterns {
    private_key: Regex,
    url_credential: Regex,
    bearer_token: Regex,
    api_token: Regex,
    typed_device_secret: Regex,
    enable_secret: Regex,
    snmp_secret: Regex,
}

static SECRET_PATTERNS: LazyLock<Result<SecretPatterns, regex::Error>> = LazyLock::new(|| {
    Ok(SecretPatterns {
        private_key: Regex::new(
            r"(?s)-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----.*?-----END [A-Z0-9 ]*PRIVATE KEY-----",
        )?,
        url_credential: Regex::new(r"(?i)\b([a-z][a-z0-9+.-]*://[^\s/:@]+:)([^\s/@]+)@")?,
        bearer_token: Regex::new(r"(?i)(Authorization:\s*Bearer\s+)([A-Za-z0-9._~+/-]+=*)")?,
        api_token: Regex::new(
            r"(?:sk-[A-Za-z0-9_-]{16,}|gh[posru]_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,}|xox[abprs]-[A-Za-z0-9-]{10,}|AKIA[0-9A-Z]{16}|ya29\.[A-Za-z0-9._-]{20,}|AIza[0-9A-Za-z_-]{30,}|glpat-[A-Za-z0-9_-]{18,})",
        )?,
        typed_device_secret: Regex::new(r"(?i)\b(password|secret)\s+([0-9])\s+(\S+)")?,
        enable_secret: Regex::new(r"(?i)\b(enable\s+secret)\s+(?:[0-9]\s+)?(\S+)")?,
        snmp_secret: Regex::new(r"(?i)\b(snmp-server\s+community)\s+(\S+)")?,
    })
});

fn marker(kind: &str) -> String {
    format!("[REDACTED:{kind}]")
}

fn replace_and_record<F>(
    text: String,
    regex: &Regex,
    kind: &str,
    matches: &mut Vec<SecretMatch>,
    replacement: F,
) -> String
where
    F: Fn(&Captures<'_>) -> String,
{
    regex
        .replace_all(&text, |captures: &Captures<'_>| {
            matches.push(SecretMatch {
                kind: kind.to_owned(),
            });
            replacement(captures)
        })
        .into_owned()
}

pub fn redact_secrets(input: &str) -> RedactedText {
    if input.is_empty() {
        return RedactedText {
            text: String::new(),
            matches: Vec::new(),
        };
    }

    let Ok(patterns) = &*SECRET_PATTERNS else {
        return RedactedText {
            text: marker("secret-scan-unavailable"),
            matches: vec![SecretMatch {
                kind: "secret-scan-unavailable".to_owned(),
            }],
        };
    };

    let mut matches = Vec::new();
    let mut text = input.to_owned();
    text = replace_and_record(
        text,
        &patterns.private_key,
        "private-key",
        &mut matches,
        |_| marker("private-key"),
    );
    text = replace_and_record(
        text,
        &patterns.url_credential,
        "url-credential",
        &mut matches,
        |caps| format!("{}{}@", &caps[1], marker("password")),
    );
    text = replace_and_record(
        text,
        &patterns.bearer_token,
        "bearer-token",
        &mut matches,
        |caps| format!("{}{}", &caps[1], marker("token")),
    );
    text = replace_and_record(text, &patterns.api_token, "api-token", &mut matches, |_| {
        marker("token")
    });
    text = replace_and_record(
        text,
        &patterns.typed_device_secret,
        "device-secret",
        &mut matches,
        |caps| format!("{} {} {}", &caps[1], &caps[2], marker("device-secret")),
    );
    text = replace_and_record(
        text,
        &patterns.enable_secret,
        "device-secret",
        &mut matches,
        |caps| format!("{} {}", &caps[1], marker("device-secret")),
    );
    text = replace_and_record(
        text,
        &patterns.snmp_secret,
        "device-secret",
        &mut matches,
        |caps| format!("{} {}", &caps[1], marker("device-secret")),
    );
    RedactedText { text, matches }
}

pub fn detect_secrets(input: &str) -> Vec<SecretMatch> {
    redact_secrets(input).matches
}

pub fn secret_kinds(input: &str) -> Vec<String> {
    detect_secrets(input)
        .into_iter()
        .map(|found| found.kind)
        .collect::<BTreeSet<_>>()
        .into_iter()
        .collect()
}

pub fn has_secret(input: &str) -> bool {
    !detect_secrets(input).is_empty()
}

pub fn redact_known_secret(input: &str, secret: Option<&str>) -> String {
    let without_known = secret.filter(|value| !value.is_empty()).map_or_else(
        || input.to_owned(),
        |secret| input.replace(secret, "[REDACTED:provider-api-key]"),
    );
    redact_secrets(&without_known).text
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn redacts_high_confidence_credentials_without_retaining_values() {
        let private_key = "-----BEGIN PRIVATE KEY-----\nabc123\n-----END PRIVATE KEY-----";
        let input = format!(
            "{private_key}\nAuthorization: Bearer abc.def-123\npostgres://alice:hunter2@db/x\npassword 7 0822455D0A16\nenable secret 5 hash\nsnmp-server community public\nsk-abcdefghijklmnop"
        );
        let redacted = redact_secrets(&input);
        for secret in [
            "abc123",
            "abc.def-123",
            "hunter2",
            "0822455D0A16",
            "hash",
            "public",
            "sk-abcdefghijklmnop",
        ] {
            assert!(!redacted.text.contains(secret));
        }
        assert!(redacted.matches.len() >= 7);
        assert!(!format!("{:?}", redacted.matches).contains("hunter2"));
    }

    #[test]
    fn avoids_low_confidence_false_positives_and_redacts_exact_provider_secret() {
        for sample in [
            "password: use your own value",
            "token budget is 4096",
            "secret sauce",
            "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQ",
        ] {
            assert_eq!(redact_secrets(sample).text, sample);
        }
        assert_eq!(
            redact_known_secret("request failed for sk-short", Some("sk-short")),
            "request failed for [REDACTED:provider-api-key]"
        );
    }
}
